src/main/scala/firrtl/graph/EulerTour.scala

```diff
@@ -0,0 +1,224 @@
+  * @param e A representation of the EulerTour as a `Seq[T]`
+  * @param h The depths of the Euler Tour represented as a `Seq[Int]`
+  */
+class EulerTour[T](r: Map[T, Int], e: Seq[T], h: Seq[Int]) {
+  private def lg(x: Double): Double = math.log(x) / math.log(2)
+
+  /** Range Minimum Query of an Euler Tour using a naive algorithm.
+    *
+    * @param x The first query bound
+    * @param y The second query bound
+    * @return The minimum between the first and second query
+    * @note The order of '''x''' and '''y''' does not matter
+    * @note '''Performance''':
+    *   - preprocessing: `O(1)`
+    *   - query: `O(n)`
+    */
+  def rmqNaive(x: T, y: T): T = {
+    val Seq(i, j) = Seq(r(x), r(y)).sorted
+    e.zip(h).slice(i, j + 1).minBy(_._2)._1
+  }
+
+  // n: the length of the Euler Tour
+  // m: the size of blocks the Euler Tour is split into
+  private val n = h.size
+  private val m = math.ceil(lg(n) / 2).toInt
+
+  /** Split up the tour into blocks of size m, padding the last block to
+    * be a multiple of m. Compute the minimum of each block, a, and
+    * the index of that minimum in each block, b.
+    */
+  private lazy val blocks = (h ++ (1 to (m - n % m))).grouped(m).toArray
+  private lazy val a = blocks map (_.min)
+  private lazy val b = blocks map (b => b.indexOf(b.min))
+
+  /** Construct a Sparse Table (ST) representation for the minimum index
+    * of a sequence of integers. Data in the returned array is indexed
+    * as: [base, power of 2 range]
+    */
+  private def constructSparseTable(x: Seq[Int]): Array[Array[Int]] = {
+    val tmp = Array.ofDim[Int](x.size + 1, math.ceil(lg(x.size)).toInt)
+    for (i <- 0 to x.size - 1; j <- 0 to math.ceil(lg(x.size)).toInt - 1) {
+      tmp(i)(j) = -1
+    }
+
+    def tableRecursive(base: Int, size: Int): Int = {
+      if (size == 0) {
+        tmp(base)(size) = base
+        base
+      } else {
+        val (a, b, c) = (base, base + (1 << (size - 1)), size - 1)
+
+        val l = if (tmp(a)(c) != -1) { tmp(a)(c)            }
+        else                         { tableRecursive(a, c) }
+
+        val r = if (tmp(b)(c) != -1) { tmp(b)(c)            }
+        else                         { tableRecursive(b, c) }
+
+        val min = if (x(l) < x(r)) l else r
+        tmp(base)(size) = min
+        assert(min >= base)
+        min
+      }
+    }
+
+    for (i <- (0 to x.size - 1);
+      j <- (0 to math.ceil(lg(x.size)).toInt - 1);
+      if i + (1 << j) - 1 < x.size) yield {
+      tableRecursive(i, j)
+    }
+    tmp
+  }
+  private lazy val st = constructSparseTable(a)
+
+  /** Precompute all possible RMQs for an array of size `n where each
+    * entry in the range is different from the last by only +-1
+    */
+  private def constructTableLookups(n: Int): Array[Array[Array[Int]]] = {
+    def sortSeqSeq[T <: Int](x: Seq[T], y: Seq[T]): Boolean = {
+      if (x(0) != y(0)) { x(0) < y(0)                }
+      else              { sortSeqSeq(x.tail, y.tail) }
```

Comment:
I think this could be written in a single line as:

```scala
if (x(0) != y(0)) x(0) < y(0) else sortSeqSeq(x.tail, y.tail)
```
I guess I don't particularly like using curly braces on if else blocks with a single statement each

Reply:
Similar idea with other if else blocks